types/underscore/underscore-tests.ts

```diff
@@ -1317,6 +1425,32 @@ declare const extractChainTypes: ChainTypeExtractor;
     extractChainTypes(_.chain(simpleString).chunk(length)); // $ExpectType ChainType<string[][], string[]>
 }
 
+// sortedIndex
+{
+    // identity iteratee
+    _.sortedIndex(simpleStringList, simpleString); // $ExpectType number
+    _(simpleStringList).sortedIndex(simpleString); // $ExpectType number
+    extractChainTypes(_.chain(simpleStringList).sortedIndex(simpleString)); // $ExpectType ChainType<number, never>
```

Comment:
Any particular reason why you went with strings for the identity iteratee while using records for the other iteratees?

Reply:
I went with strings for the identity iteratee because records are not necessarily inherently usefully comparable, and I went with records for other iteratees because the elements produced when iterating over a string do not have selectable properties that are usefully comparable (they pretty much just have `length` and that is always 1).

It would be nice to constrain this function to only work with collection and iterator combinations that are usefully comparable, but I don't think that's possible at this time because I don't think TS can differentiate between objects that have the default `toString` implementation and objects that have an overridden one.

I guess since records could override `toString` I could use records for everything, but I don't know how prevalent overriding `toString` is in the JS community (just that I never do it) or how correct it's considered to be in general, and I'd like to not set a bad example with these tests.

Let me know if you think it would be better to always compare records and assume a usefully overridden `toString` in identity iteratee tests. Conversely, let me know if overriding `toString` to make an object comparable is considered to be a major antipattern by the JS community in general that it would be useful to discourage by requiring that iteratees for this function always yield comparable primitives. Finally, let me know if I'm missing a way to assert in TS that an object type is (probably) usefully comparable because it overrides `toString` since then I could constrain iteratee results to primitives and usefully comparable objects.

Reply:
Hmm, according to [the MDN article for less than](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Less_than) it looks like possibly the correct thing to do these days to make an object usefully comparable is to add a `Symbol.toPrimitive` implementation to objects rather than overriding `toString` (though doing the latter still technically works). It looks like I could constrain objects against something like the below interface for this if you think that's a good idea.

```TS
interface Comparable {
    [Symbol.toPrimitive]: (hint: 'string' | 'number' | 'default') => string | number;
}
```

Reply:
> I went with strings for the identity iteratee because records are not necessarily inherently usefully comparable, and I went with records for other iteratees because the elements produced when iterating over a string do not have selectable properties that are usefully comparable (they pretty much just have `length` and that is always 1).

That makes perfect sense. Funny I didn't think of that.

> 
> It would be nice to constrain this function to only work with collection and iteratee combinations that are usefully comparable,

Oh no, please don't go there. That would open a big can of worms for something that people probably aren't going to do much.

> but I don't think that's possible at this time because I don't think TS can differentiate between objects that have the default `toString` implementation and objects that have an overridden one.
> 
> I guess since records could override `toString` I could use records for everything, but I don't know how prevalent overriding `toString` is in the JS community (just that I never do it) or how correct it's considered to be in general, and I'd like to not set a bad example with these tests.

This is a bit of a moot point because I'm not recommending that you try to differentiate, but: I don't know how prevalent it is in the JS community at large, either. However, I think it is perfectly correct to override the method as long as the override actually makes sense. I can also think of some authoritative overrides:

- Underscore's `_.prototype.toString`
- The JS builtin `Number.prototype.toString`(`1.toString()` returns `'1'` while `Object.prototype.toString.call(1)` returns `[object Number]`) and likewise for `Boolean`, `String`, `Symbol`, `Date` and so on

> 
> Let me know if you think it would be better to always compare records and assume a usefully overridden `toString` in identity iteratee tests.

No, I don't think so. Your tests are fine as they are.

> Conversely, let me know if overriding `toString` to make an object comparable is considered to be a major antipattern by the JS community in general that it would be useful to discourage by requiring that iteratees for this function always yield comparable primitives.

I've never seen a mention anywhere that this would be an antipattern. If I had seen it, I would have disagreed. I also don't think that a majority opinion is necessarily a reason to do something. For example, there seems to be a majority opinion in the JS community that 2-space indents are deep enough, but that is no reason for me to follow suit.

> Finally, let me know if I'm missing a way to assert in TS that an object type is (probably) usefully comparable because it overrides `toString` since then I could constrain iteratee results to primitives and usefully comparable objects.

I don't know of such a way, but again, I wouldn't be in favor of using it if I did.

Reply:
I'll leave all of this as-is then, thanks!